<a href="https://colab.research.google.com/github/maxim920/Machine-Learning/blob/main/Kosenko_lab4_1_4_7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

data = load_breast_cancer()
X = data.data
y = data.target

Перегляньте перші кілька рядків.

In [ ]:

df = pd.DataFrame(X, columns=data.feature_names)
df['target'] = y


df.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                 0.07871  ...          17.33           184.60      2019.0   
1                 0.05667  ...          23.41           158.80      1956.0   
2                 0.05999  ...          25.53           152.50      1709.0   
3                 0.09744  ...          26.50            98.87       567.7   
4                 0.05883  ...          16.67           152.20      1575.0   

   worst smoothness  worst compactness  worst concavity  worst concave points  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   worst symmetry  worst fractal dimension  target  
0          0.4601                  0.11890       0  
1          0.2750                  0.08902       0  
2          0.3613                  0.08758       0  
3          0.6638                  0.17300       0  
4          0.2364                  0.07678       0  

[5 rows x 31 columns]

Перевірка назв стовпців та типів даних

In [ ]:
print("Назви стовпців:", df.columns)

print("Типи даних:\n", df.dtypes)

Назви стовпців: Index(['mean radius', 'mean texture', 'mean perimeter', 'mean area',
       'mean smoothness', 'mean compactness', 'mean concavity',
       'mean concave points', 'mean symmetry', 'mean fractal dimension',
       'radius error', 'texture error', 'perimeter error', 'area error',
       'smoothness error', 'compactness error', 'concavity error',
       'concave points error', 'symmetry error', 'fractal dimension error',
       'worst radius', 'worst texture', 'worst perimeter', 'worst area',
       'worst smoothness', 'worst compactness', 'worst concavity',
       'worst concave points', 'worst symmetry', 'worst fractal dimension',
       'target'],
      dtype='object')
Типи даних:
 mean radius                float64
mean texture               float64
mean perimeter             float64
mean area                  float64
mean smoothness            float64
mean compactness           float64
mean concavity             float64
mean concave points        float64
mean symmetry

Перевірка наявності пропущених значень

In [ ]:

print("Пропущені значення:\n", df.isnull().sum())

Пропущені значення:
 mean radius                0
mean texture               0
mean perimeter             0
mean area                  0
mean smoothness            0
mean compactness           0
mean concavity             0
mean concave points        0
mean symmetry              0
mean fractal dimension     0
radius error               0
texture error              0
perimeter error            0
area error                 0
smoothness error           0
compactness error          0
concavity error            0
concave points error       0
symmetry error             0
fractal dimension error    0
worst radius               0
worst texture              0
worst perimeter            0
worst area                 0
worst smoothness           0
worst compactness          0
worst concavity            0
worst concave points       0
worst symmetry             0
worst fractal dimension    0
target                     0
dtype: int64


Дослідження розміру даних

In [ ]:

print("Розмір даних:", df.shape)

Розмір даних: (569, 31)


Розділення даних на навчальний та тестовий набори

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

Логістична регресія

In [ ]:
log_reg = LogisticRegression(max_iter=10000)


param_grid_log_reg = {
    'C': [0.1, 1, 10],
    'solver': ['liblinear', 'lbfgs']
}


grid_search_log_reg = GridSearchCV(log_reg, param_grid_log_reg, cv=5, scoring='accuracy')
grid_search_log_reg.fit(X_train, y_train)


print("Найкращі параметри для логістичної регресії:", grid_search_log_reg.best_params_)


y_pred_log_reg = grid_search_log_reg.predict(X_test)
print("Точність логістичної регресії:", accuracy_score(y_test, y_pred_log_reg))
print("Матриця помилок:\n", confusion_matrix(y_test, y_pred_log_reg))
print("Звіт про класифікацію:\n", classification_report(y_test, y_pred_log_reg))

Найкращі параметри для логістичної регресії: {'C': 10, 'solver': 'lbfgs'}
Точність логістичної регресії: 0.9649122807017544
Матриця помилок:
 [[ 60   3]
 [  3 105]]
Звіт про класифікацію:
               precision    recall  f1-score   support

           0       0.95      0.95      0.95        63
           1       0.97      0.97      0.97       108

    accuracy                           0.96       171
   macro avg       0.96      0.96      0.96       171
weighted avg       0.96      0.96      0.96       171



Дерево рішень

In [ ]:
dt_clf = DecisionTreeClassifier()

param_grid_dt = {
    'max_depth': [3, 5, 7],
    'min_samples_split': [2, 5, 10]
}

grid_search_dt = GridSearchCV(dt_clf, param_grid_dt, cv=5, scoring='accuracy')
grid_search_dt.fit(X_train, y_train)

print("Найкращі параметри для дерева рішень:", grid_search_dt.best_params_)

y_pred_dt = grid_search_dt.predict(X_test)
print("Точність дерева рішень:", accuracy_score(y_test, y_pred_dt))
print("Матриця помилок:\n", confusion_matrix(y_test, y_pred_dt))
print("Звіт про класифікацію:\n", classification_report(y_test, y_pred_dt))

Найкращі параметри для дерева рішень: {'max_depth': 5, 'min_samples_split': 5}
Точність дерева рішень: 0.9649122807017544
Матриця помилок:
 [[ 61   2]
 [  4 104]]
Звіт про класифікацію:
               precision    recall  f1-score   support

           0       0.94      0.97      0.95        63
           1       0.98      0.96      0.97       108

    accuracy                           0.96       171
   macro avg       0.96      0.97      0.96       171
weighted avg       0.97      0.96      0.97       171



Випадковий ліс (Random Forest)

In [ ]:
rf_clf = RandomForestClassifier()

param_grid_rf = {
    'n_estimators': [50, 100, 200],
    'max_depth': [3, 5, 7],
    'min_samples_split': [2, 5, 10]
}

grid_search_rf = GridSearchCV(rf_clf, param_grid_rf, cv=5, scoring='accuracy')
grid_search_rf.fit(X_train, y_train)

print("Найкращі параметри для випадкового лісу:", grid_search_rf.best_params_)

y_pred_rf = grid_search_rf.predict(X_test)
print("Точність випадкового лісу:", accuracy_score(y_test, y_pred_rf))
print("Матриця помилок:\n", confusion_matrix(y_test, y_pred_rf))
print("Звіт про класифікацію:\n", classification_report(y_test, y_pred_rf))

Найкращі параметри для випадкового лісу: {'max_depth': 5, 'min_samples_split': 2, 'n_estimators': 50}
Точність випадкового лісу: 0.9766081871345029
Матриця помилок:
 [[ 60   3]
 [  1 107]]
Звіт про класифікацію:
               precision    recall  f1-score   support

           0       0.98      0.95      0.97        63
           1       0.97      0.99      0.98       108

    accuracy                           0.98       171
   macro avg       0.98      0.97      0.97       171
weighted avg       0.98      0.98      0.98       171



In [ ]:
accuracy_log_reg = accuracy_score(y_test, y_pred_log_reg)
accuracy_dt = accuracy_score(y_test, y_pred_dt)
accuracy_rf = accuracy_score(y_test, y_pred_rf)

print(f"Точність логістичної регресії: {accuracy_log_reg}")
print(f"Точність дерева рішень: {accuracy_dt}")
print(f"Точність випадкового лісу: {accuracy_rf}")


best_model = None
best_accuracy = 0

if accuracy_log_reg > best_accuracy:
    best_accuracy = accuracy_log_reg
    best_model = grid_search_log_reg

if accuracy_dt > best_accuracy:
    best_accuracy = accuracy_dt
    best_model = grid_search_dt

if accuracy_rf > best_accuracy:
    best_accuracy = accuracy_rf
    best_model = grid_search_rf

print(f"Найкраща модель: {best_model.estimator.__class__.__name__} з точністю {best_accuracy}")

Точність логістичної регресії: 0.9649122807017544
Точність дерева рішень: 0.9649122807017544
Точність випадкового лісу: 0.9766081871345029
Найкраща модель: RandomForestClassifier з точністю 0.9766081871345029


In [ ]:
# Прогнозування на тестовій вибірці за допомогою найкращої моделі
y_pred_best = best_model.predict(X_test)

# Виведення результатів прогнозування
print("Результати прогнозування на тестовій вибірці:")
print("Точність:", accuracy_score(y_test, y_pred_best))
print("Матриця помилок:\n", confusion_matrix(y_test, y_pred_best))
print("Звіт про класифікацію:\n", classification_report(y_test, y_pred_best))

Результати прогнозування на тестовій вибірці:
Точність: 0.9766081871345029
Матриця помилок:
 [[ 60   3]
 [  1 107]]
Звіт про класифікацію:
               precision    recall  f1-score   support

           0       0.98      0.95      0.97        63
           1       0.97      0.99      0.98       108

    accuracy                           0.98       171
   macro avg       0.98      0.97      0.97       171
weighted avg       0.98      0.98      0.98       171



**Висновки:**
У ході лабораторної роботи було завантажено вбудований датасет load_breast_cancer з бібліотеки sklearn.datasets для задачі класифікації раку молочної залози. Проведено попередній аналіз даних, який включав перегляд перших кількох рядків, перевірку назв стовпців та типів даних, наявність пропущених значень та дослідження розміру даних. Розмір датасету склав (569, 31). Дані було розділено на навчальний і тестовий набори. Побудовано три моделі класифікації: логістична регресія, дерево рішень та випадковий ліс. Для кожної моделі було підібрано оптимальні параметри за допомогою GridSearchCV. Оцінено моделі за допомогою метрик точності (accuracy), матриці помилок (confusion matrix) та звіту про класифікацію (classification report). Усі моделі показали високу точність: логістична регресія - 0.9649, дерево рішень - 0.9649, випадковий ліс - 0.9766. На основі метрик продуктивності обрано найкращу модель - випадковий ліс (RandomForestClassifier) з точністю 0.9766. Зроблено прогноз на тестовій вибірці за допомогою найкращої моделі. Результати прогнозування показали високу точність - 0.9766, з матрицею помилок та звітом про класифікацію, що підтверджують ефективність моделі. Таким чином, у ході лабораторної роботи було успішно побудовано та оцінено моделі класифікації для задачі визначення раку молочної залози. Найкращою моделлю виявився випадковий ліс з точністю 0.9766, що свідчить про високу ефективність моделі у прогнозуванні доброякісних та злоякісних пухлин на основі характеристик клітинних ядер.